advanced_functionality/pytorch_extending_our_containers/pytorch_extending_our_containers.ipynb

```diff
@@ -0,0 +1,701 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Extending our PyTorch containers\n",
+    "\n",
+    "With Amazon SageMaker, you can package your own algorithms that can then be trained and deployed in the SageMaker environment. This notebook guides you through an example on how to extend one of our existing and predefined SageMaker deep learning framework containers.\n",
+    "\n",
+    "By packaging an algorithm in a container, you can bring almost any code to the Amazon SageMaker environment, regardless of programming language, environment, framework, or dependencies. \n",
+    "\n",
+    "1. [Extending our PyTorch containers](#Extending-our-pytorch-containers)\n",
+    "  1. [When should I extend a SageMaker container?](#When-should-I-extend-a-SageMaker-container?)\n",
+    "  1. [Permissions](#Permissions)\n",
+    "  1. [The example](#The-example)\n",
+    "  1. [The presentation](#The-presentation)\n",
+    "1. [Part 1: Packaging and Uploading your Algorithm for use with Amazon SageMaker](#Part-1:-Packaging-and-Uploading-your-Algorithm-for-use-with-Amazon-SageMaker)\n",
+    "    1. [An overview of Docker](#An-overview-of-Docker)\n",
+    "    1. [How Amazon SageMaker runs your Docker container](#How-Amazon-SageMaker-runs-your-Docker-container)\n",
+    "      1. [Running your container during training](#Running-your-container-during-training)\n",
+    "        1. [The input](#The-input)\n",
+    "        1. [The output](#The-output)\n",
+    "      1. [Running your container during hosting](#Running-your-container-during-hosting)\n",
+    "    1. [The parts of the sample container](#The-parts-of-the-sample-container)\n",
+    "    1. [The Dockerfile](#The-Dockerfile)\n",
+    "    1. [Building and registering the container](#Building-and-registering-the-container)\n",
+    "  1. [Testing your algorithm on your local machine](#Testing-your-algorithm-on-your-local-machine)\n",
+    "  1. [Download the CIFAR-10 dataset](#Download-the-CIFAR-10-dataset)\n",
+    "  1. [SageMaker Python SDK Local Training](#SageMaker-Python-SDK-Local-Training)\n",
+    "  1. [Fit, Deploy, Predict](#Fit,-Deploy,-Predict)\n",
+    "  1. [Making predictions using Python SDK](#Making-predictions-using-Python-SDK)\n",
+    "1. [Part 2: Training and Hosting your Algorithm in Amazon SageMaker](#Part-2:-Training-and-Hosting-your-Algorithm-in-Amazon-SageMaker)\n",
+    "  1. [Set up the environment](#Set-up-the-environment)\n",
+    "  1. [Create the session](#Create-the-session)\n",
+    "  1. [Upload the data for training](#Upload-the-data-for-training)\n",
+    "  1. [Training On SageMaker](#Training-on-SageMaker)\n",
+    "  1. [Optional cleanup](#Optional-cleanup)  \n",
+    "1. [Reference](#Reference)\n",
+    "\n",
+    "_or_ I'm impatient, just [let me see the code](#The-Dockerfile)!\n",
+    "\n",
+    "## When should I extend a SageMaker container?\n",
```

Comment:
Similar to the "When should I build my own" section of the [TensorFlow notebook](https://github.com/awslabs/amazon-sagemaker-examples/blob/master/advanced_functionality/tensorflow_bring_your_own/tensorflow_bring_your_own.ipynb) can we add some specific, concrete examples for what type of functionality users might want to add or what configurations they might want to change?

Reply:
Yes, I added it in but it... isn't bolded or in its own section.

'Even if there is direct SDK support for your environment or framework, you may want to add additional functionality or configure your container environment differently while utilizing our container to use on SageMaker.'

There really is only one reason to extend our containers, and that is to add additional libraries or modify configurations.

Reply:
I worry "add additional functionality or configure your container" is too broad.  Specific examples.  E.g. "If you have a specific Python library like Pillow that we don't currently install but you always want included in your container."  etc.  This is kind of a minor point though.